File: app/app/src/main/java/me/williamhester/reddit/ui/activities/ContentActivity.java
package me.williamhester.reddit.ui.activities;

import android.os.Bundle;
import android.support.design.widget.BaseTransientBottomBar;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;

import org.greenrobot.eventbus.Subscribe;

import me.williamhester.reddit.R;
import me.williamhester.reddit.messages.FailedRedditRequestMessage;
import me.williamhester.reddit.models.Submission;
import me.williamhester.reddit.ui.fragments.CommentsFragment;

/** Activity that holds basic content. */
public class ContentActivity extends BaseActivity {

  public static final String TYPE_EXTRA = "type";
  public static final String VOTABLE_EXTRA = "votable";
  public static final String PERMALINK_EXTRA = "permalink";

  public static final String COMMENTS = "comments";

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    Fragment f = getSupportFragmentManager().findFragmentById(R.id.fragment_container);
    if (f == null) {
      getSupportFragmentManager().beginTransaction()
          .add(R.id.fragment_container, createContentFragment())
          .commit();
    }
  }

  @Override
  protected int getLayoutId() {
    return R.layout.activity_content;
  }

  protected Fragment createContentFragment() {
    Bundle args = getIntent().getExtras();
    String type = args.getString(TYPE_EXTRA);
    if (type == null) {
      return null;
    }
    switch (type) {
      case COMMENTS:
        String permalink = args.getString(PERMALINK_EXTRA);
        Submission s = args.getParcelable(VOTABLE_EXTRA);
        return CommentsFragment.newInstance(permalink, s);
      default:
        return null;
    }
  }

  @Subscribe
  public void onRequestFailure(FailedRedditRequestMessage message) {
    Snackbar
        .make(
            findViewById(R.id.fragment_container),
            R.string.request_failed,
            BaseTransientBottomBar.LENGTH_SHORT)
        .show();
  }
}

File: app/app/src/main/java/me/williamhester/reddit/ui/activities/ContentActivity.kt
package me.williamhester.reddit.ui.activities

import android.os.Bundle
import androidx.fragment.app.Fragment
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import me.williamhester.reddit.R
import me.williamhester.reddit.messages.FailedRedditRequestMessage
import me.williamhester.reddit.models.Submission
import me.williamhester.reddit.ui.fragments.CommentsFragment
import org.greenrobot.eventbus.Subscribe

/** Activity that holds basic content.  */
open class ContentActivity : BaseActivity() {

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    val f = supportFragmentManager.findFragmentById(R.id.fragment_container)
    if (f == null) {
      supportFragmentManager.beginTransaction()
          .add(R.id.fragment_container, createContentFragment()!!)
          .commit()
    }
  }

  override val layoutId: Int
    get() = R.layout.activity_content

  protected open fun createContentFragment(): Fragment? {
    val args = intent.extras
    val type = args?.getString(TYPE_EXTRA) ?: return null
    return when (type) {
      COMMENTS -> {
        val permalink = args.getString(PERMALINK_EXTRA)!!
        val s = args.getParcelable<Submission>(VOTABLE_EXTRA)!!
        CommentsFragment.newInstance(permalink, s)
      }
      else -> null
    }
  }

  @Subscribe
  fun onRequestFailure(message: FailedRedditRequestMessage) {
    Snackbar
        .make(
            findViewById(R.id.fragment_container),
            R.string.request_failed,
            BaseTransientBottomBar.LENGTH_SHORT)
        .show()
  }

  companion object {
    const val TYPE_EXTRA = "type"
    const val VOTABLE_EXTRA = "votable"
    const val PERMALINK_EXTRA = "permalink"
    const val COMMENTS = "comments"
  }
}
